File: implementations/java-springboot/examples/src/main/java/com/example/accountservice/account/application/event/MoneyWithdrawnEventHandler.java
package com.example.accountservice.account.application.event;

import com.example.accountservice.account.application.service.NotificationService;
import com.example.accountservice.account.domain.MoneyWithdrawnEvent;
import com.example.accountservice.outbox.OutboxEventHandler;
import com.fasterxml.jackson.databind.ObjectMapper;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Component;

/**
 * Handles the {@link MoneyWithdrawnEvent} accumulated in the Outbox to send a notification email to
 * the account owner.
 */
@Component
@RequiredArgsConstructor
public class MoneyWithdrawnEventHandler implements OutboxEventHandler {

    private final NotificationService notificationService;
    private final ObjectMapper objectMapper;

    @Override
    public String eventType() {
        return MoneyWithdrawnEvent.class.getSimpleName();
    }

    @Override
    public void handle(String payload) throws Exception {
        MoneyWithdrawnEvent event = objectMapper.readValue(payload, MoneyWithdrawnEvent.class);
        notificationService.sendEmail(
                event.accountId(),
                "MoneyWithdrawn",
                event.email(),
                "[Account] Your withdrawal is complete",
                event.amount().amount()
                        + " "
                        + event.amount().currency()
                        + " has been withdrawn. Balance after withdrawal: "
                        + event.balanceAfter().amount()
                        + " "
                        + event.balanceAfter().currency());
    }
}

File: implementations/kotlin-springboot/examples/src/main/kotlin/com/example/accountservice/account/application/event/MoneyWithdrawnEventHandler.kt
package com.example.accountservice.account.application.event

import com.example.accountservice.account.domain.MoneyWithdrawnEvent
import com.example.accountservice.notification.application.service.NotificationService
import org.springframework.stereotype.Component

/**
 * Sends a withdrawal-completed notification email when MoneyWithdrawnEvent is delivered via the Outbox.
 *
 * Exceptions are not caught and are thrown as-is — [com.example.accountservice.outbox.OutboxConsumer]
 * catches them, logs, and leaves the Outbox row as processed=false so the next call retries it
 * (at-least-once delivery).
 *
 * [eventId] is the `eventId` of the Outbox row this event was stored in — [NotificationService] uses it
 * as the key to apply Level 2 (Ledger) duplicate-send prevention.
 */
@Component
class MoneyWithdrawnEventHandler(
    private val notificationService: NotificationService,
) {
    fun handle(
        event: MoneyWithdrawnEvent,
        eventId: String,
    ) {
        notificationService.sendEmail(
            accountId = event.accountId,
            eventType = "MoneyWithdrawn",
            sourceEventId = eventId,
            recipient = event.email,
            subject = "[Account] Your withdrawal is complete",
            body =
                "${event.amount.amount} ${event.amount.currency} has been withdrawn. " +
                    "Balance after withdrawal: ${event.balanceAfter.amount} ${event.balanceAfter.currency}",
        )
    }
}
